scipy/spatial/transform/_rotation.py

```diff
@@ -2575,43 +1559,14 @@ def apply(self, vectors, inverse=False):
                [ 1.09533535, -0.8365163 ,  0.3169873 ]])
 
         """
-        vectors = np.asarray(vectors)
-        if vectors.ndim > 2 or vectors.shape[-1] != 3:
-            raise ValueError("Expected input of shape (3,) or (P, 3), "
-                             "got {}.".format(vectors.shape))
-
-        cdef bint single_vector = vectors.ndim == 1
-        cdef Py_ssize_t n_vectors = 1 if single_vector else len(vectors)
-        cdef Py_ssize_t n_rotations = 1 if self.single else len(self)
-
-        if n_vectors != 1 and n_rotations != 1 and n_vectors != n_rotations:
-            raise ValueError("Expected equal numbers of rotations and vectors "
-                             ", or a single rotation, or a single vector, got "
-                             "{} rotations and {} vectors.".format(
-                                n_rotations, n_vectors))
-
-        cdef np.ndarray matrix = self.as_matrix()
-
-        if inverse:
-            matrix = np.swapaxes(matrix, -1, -2)
-
-        if single_vector:
-            return np.matmul(matrix, vectors)
-
-        if self.single:
-            matrix = matrix[None, :, :]
-
-        if n_rotations == 1:
-            # Single rotation/many vectors, use matmul for speed: The axes argument
-            # is such that the input arguments don't need to be transposed and the
-            # output argument is contineous in memory.
-            return np.matmul(matrix, vectors, axes=[(-2, -1), (-1, -2), (-1, -2)])[0]
```

Comment:
This fast path was removed, does einsum show equal in speed now?

Reply:
Not to my knowledge, this might have been by accident. Will double-check

Reply:
After checking this function again it seems like a lot has changed between the old implementation and the one that was added here, not sure why. The last commit has revised the function, so it would be great if you could have another look at it. This one should have the same optimizations.